src/category_theory/monad/equiv_mon.lean

```diff
@@ -0,0 +1,233 @@
+      refl,
+    end } }
+variable {C}
+
+/-- To every monoid object in `C ⥤ C` we associate a `Monad C`. -/
+def of_Mon : Mon_ (C ⥤ C) → Monad C := λ M,
+{ func := M.X,
+  str :=
+  { η := M.one,
+    μ := M.mul,
+    assoc' := begin
+      intro X,
+      rw ←nat_trans.hcomp_id_app,
+      change ((M.mul ◫ 𝟙 M.X) ≫ M.mul).app X = _,
+      erw M.mul_assoc,
+      simp only [nat_trans.comp_app],
+      change ((α_ M.X M.X M.X).app X).hom ≫ (_ ◫ _).app X ≫ _ = _,
+      suffices : ((α_ M.X M.X M.X).app X).hom = 𝟙 _, by {rw this, simp},
+      refl,
+    end,
+    left_unit' := begin
+      intro X,
+      have := M.mul_one,
+      change (_ ◫ _) ≫ _ = _ at this,
+      rw [←nat_trans.id_hcomp_app, ←nat_trans.comp_app, this],
+      refl,
+    end,
+    right_unit' := begin
+      intro X,
+      have := M.one_mul,
+      change (_ ◫ _) ≫ _ = _ at this,
+      rw [←nat_trans.hcomp_id_app, ←nat_trans.comp_app, this],
+      refl,
+    end } }
+
+variable (C)
+/-- Passing from `Mon_ (C ⥤ C)` to `Monad C` is functorial. -/
+def Mon_to_Monad : Mon_ (C ⥤ C) ⥤ Monad C :=
+{ obj := of_Mon,
+  map := λ M N f,
+  { app_η' := begin
+      intro X,
+      simp only [auto_param_eq],
+      erw [←nat_trans.comp_app,f.one_hom],
+      refl,
+    end,
+    app_μ' := begin
+      intro X,
+      simp only [auto_param_eq],
+      erw [←nat_trans.comp_app, f.mul_hom],
+      simp only [nat_trans.naturality, assoc, nat_trans.comp_app],
+      erw [nat_trans.hcomp_app, assoc],
+      refl,
+    end,
+    ..f.hom } }
+variable {C}
+
+theorem of_to_mon_end_obj (M : Mon_ (C ⥤ C)) : (Mon_to_Monad C ⋙ Monad_to_Mon C).obj M = M :=
+  by {apply Mon_.hext, repeat {refl}}
+
+theorem to_of_mon_end_obj (M : Monad C) : (Monad_to_Mon C ⋙ Mon_to_Monad C).obj M = M :=
+  by {apply Monad.hext, repeat {refl}}
+
+theorem of_to_mon_end : Mon_to_Monad C ⋙ Monad_to_Mon C = 𝟭 _ :=
+begin
+  apply functor.ext,
+  { intros X Y f,
+    ext,
+    simp only [functor.id_map, functor.comp_map, Mon_.comp_hom', nat_trans.comp_app,
+      Mon_.hom_eq_to_hom, eq_to_hom_app, id_comp, eq_to_hom_refl, comp_id],
+    refl },
+  { intro X,
+    apply of_to_mon_end_obj },
+end
+
+theorem to_of_mon_end : Monad_to_Mon C ⋙ Mon_to_Monad _ = 𝟭 _ :=
+begin
+  apply functor.ext,
+  { intros X Y f,
+    ext,
+    simp only [Monad.comp_to_nat_trans, functor.id_map, functor.comp_map, nat_trans.vcomp_eq_comp,
+      nat_trans.comp_app, to_nat_trans_eq_to_hom, eq_to_hom_app, id_comp, eq_to_hom_refl, comp_id],
+    refl },
+  { intro X,
+    apply to_of_mon_end_obj }
+end
+
+/-- Isomorphism of functors used in `Monad_Mon_equiv` -/
+def of_to_mon_end_iso : Mon_to_Monad C ⋙ Monad_to_Mon C ≅ 𝟭 _ :=
+{ hom :=
+  { app := λ M,
+  { hom := 𝟙 _ } },
+  inv :=
+  { app := λ M,
+  { hom := 𝟙 _ } } }
+
+/-- Isomorphism of functors used in `Monad_Mon_equiv` -/
+def to_of_mon_end_iso : Monad_to_Mon C ⋙ Mon_to_Monad C ≅ 𝟭 _ :=
+{ hom :=
+  { app := λ M,
+  { app := λ X, 𝟙 _,
+    app_η' := by tidy,
+    app_μ' := by tidy } },
```

Comment:
You can probably omit these as the `auto_param` will handle them.

Reply:
I had the same thought originally, but it didn't work for some reason. Did I declare the `auto_param` incorrectly in the definition of `monad_hom`?

Reply:
Okay, I figured it out. When defining monad_hom I had fields of the form
`(app_η' {X} : ... . obviously)`
I changed it to 
`(app_η' : ∀ {X}, ... . obviously)`
and now the `auto_param` works as expected. Is this expected behavior?

Reply:
No, I hadn't known about this issue. Can you distill it down to a minimal example? It seems worth reporting as an issue.